src/Kiota.Builder/Refiners/GoRefiner.cs

```diff
@@ -218,6 +219,192 @@ public override Task RefineAsync(CodeNamespace generatedCode, CancellationToken
+        foreach (var cycle in cycles)
+        {
+            foreach (var cycleReference in cycle.Value)
+            {
+                var dupNS = cycleReference[^2]; // 2nd last element is target base namespace
+                var nameSpace = modelsNameSpace.FindNamespaceByName(dupNS, true);
+
+                migratedNamespaces[dupNS] = modelsNameSpace.Name;
+                MigrateNameSpace(nameSpace!, modelsNameSpace);
+
+                if (!cycle.Key.Equals(modelsNameSpace.Name, StringComparison.OrdinalIgnoreCase) && !migratedNamespaces.ContainsKey(cycle.Key))
+                {
+                    migratedNamespaces[cycle.Key] = modelsNameSpace.Name;
+                    MigrateNameSpace(modelsNameSpace.FindNamespaceByName(cycle.Key, true)!, modelsNameSpace);
+                }
+            }
+        }
+
+        CorrectReferencesToMigratedModels(currentElement, migratedNamespaces);
+    }
+
+    private string GetComposedName(CodeElement codeClass)
+    {
+        var classNameList = getPathsName(codeClass, codeClass.Name);
+        return string.Join(string.Empty, classNameList.Count > 1 ? classNameList.Skip(1) : classNameList);
+    }
+
+    private void GetUsingsInModelsNameSpace(CodeNamespace modelsNameSpace, CodeNamespace currentNameSpace, Dictionary<string, HashSet<string>> dependencies)
+    {
+        if (!modelsNameSpace.Name.Equals(currentNameSpace.Name, StringComparison.OrdinalIgnoreCase) && !currentNameSpace.IsChildOf(modelsNameSpace))
+            return;
+
+        dependencies[currentNameSpace.Name] = currentNameSpace.Classes
+            .SelectMany(codeClass => codeClass.Usings)
+            .Where(x => x.Declaration != null && !x.Declaration.IsExternal)
+            .Select(static x => x.Declaration?.TypeDefinition?.GetImmediateParentOfType<CodeNamespace>())
+            .Where(ns => ns != null && ns.Name != currentNameSpace.Name)
+            .Select(static ns => ns!.Name)
+            .ToHashSet();
+
+        currentNameSpace.Namespaces
+            .Where(codeNameSpace => !dependencies.ContainsKey(codeNameSpace.Name))
+            .ToList()
+            .ForEach(codeNameSpace => GetUsingsInModelsNameSpace(modelsNameSpace, codeNameSpace, dependencies));
+    }
+
+    /// <summary>
+    /// Returns a dictionary of cycles in the graph with the key being the base namespace and the values being the path to the cycle
+    /// In GoLang, any self referencing namespace in a tree is a cycle, therefore the whole namespace is moved to the root
+    /// </summary>
+    /// <param name="dependencies"></param>
+    /// <returns></returns>
+    static Dictionary<string, List<List<string>>> FindCycles(Dictionary<string, HashSet<string>> dependencies)
+    {
+        var cycles = new Dictionary<string, List<List<string>>>();
+        var visited = new HashSet<string>();
+        var stack = new Stack<string>();
+
+        foreach (var node in dependencies.Keys)
+        {
+            if (!visited.Contains(node))
```

Comment:
isn't this going to be sensitive to the order in which the code dom is?

Reply:
traversal of the graph would be affected by the order of the dom but to address this and ensure the results are always the same, whenever a cycle is detected, both namespace of the models with a cycle are collapsed to the root of the model namespace.